lib/view_model/active_record/cache.rb

```diff
@@ -35,85 +53,98 @@ def clear
-    missing_ids = ids.to_set.subtract(cached_serializations.keys)
+    def render_from_cache(viewmodel_class, ids, initial_viewmodels: nil, locked: false)
+      viewmodel_class.transaction do
+        root_serializations = Array.new(ids.size)
 
-    # If initial viewmodels have been locked, we can serialize them for cache
-    #  misses.
-    available_viewmodels =
-      if locked
-        initial_viewmodels&.each_with_object({}) do |vm, h|
-          h[vm.id] = vm if missing_ids.include?(vm.id)
+        # Collect input array positions for each id, allowing duplicates
+        positions = ids.each_with_index.with_object({}) do |(id, i), h|
+          (h[id] ||= []) << i
         end
-      end
 
-    @viewmodel_class.transaction do
-      # Load remaining views and serialize
-      viewmodels = worker.find_and_preload_viewmodels(@viewmodel_class, missing_ids.to_a,
-                                                      available_viewmodels: available_viewmodels)
+        # If duplicates are specified, fetch each only once
+        ids = positions.keys
+
+        ids_to_render = ids.to_set
 
-      loaded_serializations = worker.serialize_and_cache(viewmodels)
-      loaded_serializations.each do |id, data|
-        positions[id].each do |idx|
-          data_serializations[idx] = data
+        if viewmodel_class < CacheableView
+          # Load existing serializations from the cache
+          cached_serializations = load_from_cache(viewmodel_class.viewmodel_cache, ids)
+          cached_serializations.each do |id, data|
+            positions[id].each do |idx|
+              root_serializations[idx] = data
+            end
+          end
+
+          ids_to_render.subtract(cached_serializations.keys)
+
+          # If initial root viewmodels were provided, call hooks on any
+          # viewmodels which were rendered from the cache to ensure that the
+          # root is visible (in isolation). Other than this, no traversal
+          # callbacks are performed for cache-rendered views. This particularly
+          # requires care for references: if a visible view may refer to
+          # non-visible cacheable views, those referenced views will not be
+          # access control checked.
+          initial_viewmodels&.each do |v|
+            next unless cached_serializations.has_key?(v.id)
+            serialize_context.run_callback(ViewModel::Callbacks::Hook::BeforeVisit, v)
+            serialize_context.run_callback(ViewModel::Callbacks::Hook::AfterVisit, v)
+          end
         end
-      end
 
-      # Resolve references
-      worker.resolve_references!
+        # Render remaining views. If initial viewmodels have been locked, we may
+        # use them to serialize from, otherwise we must reload with share lock
+        # in find_and_preload.
+        available_viewmodels =
+          if locked
+            initial_viewmodels&.each_with_object({}) do |vm, h|
+              h[vm.id] = vm if missing_ids.include?(vm.id)
```

Comment:
I don't see where `missing_ids` is defined. It looks like it should be using `ids_to_render`. Missing test case?

Reply:
Yeah, that's a missed rename from the refactor, and absolutely a missing test case.

Reply:
fixed and added test cases